src/main/kotlin/uk/gov/justice/digital/hmpps/prisonertonomisupdate/casenotes/CaseNotesReconciliationService.kt

```diff
@@ -0,0 +1,207 @@
+import java.time.format.DateTimeFormatter
+import kotlin.String
+
+@Service
+class CaseNotesReconciliationService(
+  private val telemetryClient: TelemetryClient,
+  private val caseNotesDpsApiService: CaseNotesDpsApiService,
+  private val caseNotesNomisApiService: CaseNotesNomisApiService,
+  private val nomisApiService: NomisApiService,
+  @Value("\${reports.casenotes.reconciliation.page-size}")
+  private val pageSize: Long = 20,
+) {
+  internal companion object {
+    val log: Logger = LoggerFactory.getLogger(this::class.java)
+  }
+
+  suspend fun generateReconciliationReport(prisonerCount: Long): List<MismatchCaseNote> =
+    prisonerCount.asPages(pageSize).flatMap { page ->
+      val prisoners = getAllPrisonersForPage(page)
+
+      withContext(Dispatchers.Unconfined) {
+        prisoners.map { async { checkMatch(it) } }
+      }.awaitAll().filterNotNull()
+    }
+
+  internal suspend fun getAllPrisonersForPage(page: Pair<Long, Long>) =
+    runCatching { nomisApiService.getAllPrisoners(page.first, page.second).content }
+      .onFailure {
+        telemetryClient.trackEvent(
+          "casenotes-reports-reconciliation-mismatch-page-error",
+          mapOf(
+            "page" to page.first.toString(),
+          ),
+        )
+        log.error("Unable to match entire page of prisoners: $page", it)
+      }
+      .getOrElse { emptyList() }
+      .also { log.info("Page requested: $page, with ${it.size} prisoners") }
+
+  suspend fun checkMatch(prisonerId: PrisonerId): MismatchCaseNote? {
+    val offenderNo = prisonerId.offenderNo
+    return runCatching {
+      val nomisCaseNotes = caseNotesNomisApiService.getCaseNotesForPrisoner(offenderNo)
+        .caseNotes
+        .map { c ->
+          CommonCaseNoteFields(
+            c.caseNoteText,
+            c.caseNoteType.code,
+            c.caseNoteSubType.code,
+            c.occurrenceDateTime,
+            c.authorUsername,
+            c.amendments.map { a ->
+              CommonAmendmentFields(
+                text = a.text,
+                occurrenceDateTime = a.createdDateTime,
+                authorUsername = a.authorUsername,
+              )
+            },
+          )
+        }
+        .toSortedSet(fieldsComparator)
+
+      val dpsCaseNotes = caseNotesDpsApiService.getCaseNotesForPrisoner(offenderNo)
+        .filter { !it.sensitive }
+        .map { c ->
+          CommonCaseNoteFields(
+            c.text,
+            c.type,
+            c.subType,
+            c.occurrenceDateTime.format(DateTimeFormatter.ISO_DATE_TIME),
+            c.authorUsername,
+            c.amendments.map { a ->
+              CommonAmendmentFields(
+                text = a.additionalNoteText,
+                occurrenceDateTime = a.creationDateTime?.format(DateTimeFormatter.ISO_DATE_TIME),
+                authorUsername = a.authorUserName,
+              )
+            },
+          )
+        }
+        .toSortedSet(fieldsComparator)
```

Comment:
the method is too long - next time i might refactor e.g. the mappings to CommonCaseNoteFields into transform methods